OpenRA.Mods.Common/Traits/World/BuildableTerrainOverlay.cs

```diff
@@ -0,0 +1,83 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	class BuildableTerrainOverlayInfo : TraitInfo
+	{
+		[FieldLoader.Require]
+		[ActorReference(typeof(BuildingInfo))]
+		public readonly string CheckOnBuilding = null;
+
+		[PaletteReference]
+		[Desc("Palette to use for rendering the sprite.")]
+		public readonly string Palette = TileSet.TerrainPaletteInternalName;
+
+		[Desc("Sprite definition.")]
+		public readonly string Image = "overlay";
+
+		[SequenceReference("Image")]
+		[Desc("Sequence to use for unbuildable area.")]
+		public readonly string Sequence = "build-invalid";
+
+		public override object Create(ActorInitializer init)
+		{
+			return new BuildableTerrainOverlay(init.Self, this);
+		}
+	}
+
+	class BuildableTerrainOverlay : IRenderAboveShroud, IWorldLoaded
+	{
+		readonly BuildableTerrainOverlayInfo info;
+		readonly Sprite disabledSprite;
+		readonly HashSet<string> allowedTerrain;
+
+		public bool Enabled = false;
+
+		PaletteReference palette;
+
+		public BuildableTerrainOverlay(Actor self, BuildableTerrainOverlayInfo info)
+		{
+			this.info = info;
+
+			var rules = self.World.Map.Rules;
+			disabledSprite = rules.Sequences.GetSequence(info.Image, info.Sequence).GetSprite(0);
+			allowedTerrain = rules.Actors[info.CheckOnBuilding.ToLowerInvariant()].TraitInfo<BuildingInfo>().TerrainTypes;
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		{
+			palette = wr.Palette(info.Palette);
+		}
+
+		public IEnumerable<IRenderable> RenderAboveShroud(Actor self, WorldRenderer wr)
+		{
+			if (info.CheckOnBuilding == "" || !self.World.Map.Rules.Actors.ContainsKey(info.CheckOnBuilding) || !Enabled)
+				yield break;
+
+			var cells = self.World.Map.AllCells.Where(c => self.World.Map.Contains(c)
```

Comment:
Evaluating this every render frame is going to be awful for performance. Use a `TerrainSpriteLayer` and bind to the tiles/customterrain changing instead.

Reply:
I rewrote it using `TerrainSpriteLayer` though that regression isometric mods.